File: 55-jump-game/jump-game.cpp
class Solution {
public:
    inline bool canJump(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return true;

        int i = 0, maxReachable = 0;
        while (i < n) {
            maxReachable = max(maxReachable, i + nums[i]);
            if (i < maxReachable) {
                i += 1;
            } else {
                break;
            }
        }

        if (i >= n-1) {
            return true;
        }

        return false;
    }
};

File: 55-jump-game/jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int maxReachability = 0;
        int i = 0;
        while (i < n) {
            maxReachability = Math.max(maxReachability, i + nums[i]);
            if (i < maxReachability) {
                i += 1;
            } else {
                break;
            }
        }

        return i >= n-1;
    }
}